# <u>**Streaming Service Comparison**</u>

### **Objective:**
- Determine which streaming platform hosts the majority of content I enjoy so that I can pare down the services to which I subscribe. 

### **Data Sources:**
- [Netflix via Kaggle.com](https://www.kaggle.com/datasets/octopusteam/full-netflix-dataset)
- [Hulu via Kaggle.com](https://www.kaggle.com/datasets/octopusteam/full-hulu-dataset)
- [Prime via Kaggle.com](https://www.kaggle.com/datasets/octopusteam/full-amazon-prime-dataset/data)
- [AppleTV via Kaggle.com](https://www.kaggle.com/datasets/octopusteam/full-apple-tv-dataset)

## **Data Collection & Loading**

### **Import Pandas, Numpy, Matplotlib, Wordcloud, and PIL**

In [1]:
import glob
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import wordcloud as wc
from PIL import Image

### **Data Load**

In [2]:
# Load in each file separately. 
# Plan to create a function for this or NOT
apple = pd.read_csv("AppleTV.csv")
hulu = pd.read_csv("Hulu.csv")
netflix = pd.read_csv("Netflix.csv")
prime = pd.read_csv("Prime.csv")

### **Apple**   

In [3]:
#display(apple)
# apple.head(3)
# hulu.head(3)
# netflix.head(3)
# prime.head(3)
#display(apple)
print(apple.head(3))
print(hulu.head(3))
print(netflix.head(3))
print(prime.head(3))

             title   type          genres  releaseYear     imdbId  \
0       Four Rooms  movie          Comedy       1995.0  tt0113101   
1     Forrest Gump  movie  Drama, Romance       1994.0  tt0109830   
2  American Beauty  movie           Drama       1999.0  tt0169547   

   imdbAverageRating  imdbNumVotes availableCountries  
0                6.7      113403.0                NaN  
1                8.8     2348885.0                NaN  
2                8.3     1238903.0                NaN  
                 title   type                  genres  releaseYear     imdbId  \
0                Ariel  movie  Comedy, Crime, Romance       1988.0  tt0094675   
1  Shadows in Paradise  movie    Comedy, Drama, Music       1986.0  tt0092149   
2         Forrest Gump  movie          Drama, Romance       1994.0  tt0109830   

   imdbAverageRating  imdbNumVotes availableCountries  
0                7.4        8949.0                NaN  
1                7.5        7727.0                NaN  
2     

In [4]:
# Dispaly information about the dataset (i.e., row counts, column counts, column names, datatypes, # of non-null rows)
apple.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18208 entries, 0 to 18207
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               17643 non-null  object 
 1   type                18208 non-null  object 
 2   genres              17549 non-null  object 
 3   releaseYear         18179 non-null  float64
 4   imdbId              16727 non-null  object 
 5   imdbAverageRating   16314 non-null  float64
 6   imdbNumVotes        16314 non-null  float64
 7   availableCountries  84 non-null     object 
dtypes: float64(3), object(5)
memory usage: 1.1+ MB


In [5]:
apple.describe()

releaseYear  imdbAverageRating  imdbNumVotes
count  18179.000000       16314.000000  1.631400e+04
mean    2006.822432           6.381574  2.554855e+04
std       18.410460           1.162677  1.008288e+05
min     1902.000000           1.300000  5.000000e+00
25%     2001.000000           5.700000  2.010000e+02
50%     2014.000000           6.500000  1.234000e+03
75%     2020.000000           7.200000  8.106500e+03
max     2025.000000           9.600000  2.348885e+06

In [6]:
apple.select_dtypes("object").describe()

title   type genres     imdbId  \
count                          17643  18208  17549      16727   
unique                         16986      2    798      16724   
top     Teenage Mutant Ninja Turtles  movie  Drama  tt0112120   
freq                               5  14005   1590          2   

       availableCountries  
count                  84  
unique                 14  
top                    US  
freq                   42

In [7]:
# Drop availableCountries Column - no value with such a low non-null number
apple.drop("availableCountries",axis=1,inplace=True)
apple.columns

Index(['title', 'type', 'genres', 'releaseYear', 'imdbId', 'imdbAverageRating',
       'imdbNumVotes'],
      dtype='object')

In [8]:
# Change releaseYear,  imdbNumVotes to integers 
# Display only the data types

apple[["releaseYear","imdbNumVotes"]] = apple[["releaseYear","imdbNumVotes"]].apply(np.int64)
display(apple.dtypes)

c:\Users\rkynhoff\OneDrive - University of Rochester\Documents\CodeLouisville\Streaming_Service_Comparisons\venv\Lib\site-packages\pandas\core\series.py:1031: RuntimeWarning: invalid value encountered in cast
  arr = np.asarray(values, dtype=dtype)


title                 object
type                  object
genres                object
releaseYear            int64
imdbId                object
imdbAverageRating    float64
imdbNumVotes           int64
dtype: object

In [9]:
# Rename Columns
apple.rename(columns={"title":"Title","type":"Type","genres":"Genres","releaseYear":"Release Year","imdbId":"IMDb ID","imdbAverageRating":"IMDb Average Rating","imdbNumVotes":"IMDb Num Votes"},inplace = True)
# Add a column to hulu DF and entered a default value for each row & display bottom two rows
apple["Service Name"] = "AppleTV"
apple.columns

Index(['Title', 'Type', 'Genres', 'Release Year', 'IMDb ID',
       'IMDb Average Rating', 'IMDb Num Votes', 'Service Name'],
      dtype='object')

In [10]:
# Determine how many titles contain null values
apple["Title"].isna().sum()

np.int64(565)

In [11]:
# Drop the rows contiaining null values in the Title column
apple = apple.dropna(subset=["Title"])
apple["Title"].isna().sum()


np.int64(0)

In [12]:
apple.duplicated().value_counts()

False    17641
True         2
Name: count, dtype: int64

In [13]:
apple[apple.duplicated(keep=False)]

Title   Type  \
3497   A Personal Journey with Martin Scorsese Throug...  movie   
13091                The Kidnap & Murder of Lynda Spence  movie   
13093                The Kidnap & Murder of Lynda Spence  movie   
13980  A Personal Journey with Martin Scorsese Throug...  movie   

                                Genres  Release Year    IMDb ID  \
3497   Biography, Documentary, History          1995  tt0112120   
13091               Documentary, Crime          2023        NaN   
13093               Documentary, Crime          2023        NaN   
13980  Biography, Documentary, History          1995  tt0112120   

       IMDb Average Rating       IMDb Num Votes Service Name  
3497                   8.5                 5196      AppleTV  
13091                  NaN -9223372036854775808      AppleTV  
13093                  NaN -9223372036854775808      AppleTV  
13980                  8.5                 5196      AppleTV

In [14]:
apple.drop_duplicates()


Title   Type              Genres  \
0                                Four Rooms  movie              Comedy   
1                              Forrest Gump  movie      Drama, Romance   
2                           American Beauty  movie               Drama   
3                              Citizen Kane  movie      Drama, Mystery   
4                                Metropolis  movie       Drama, Sci-Fi   
...                                     ...    ...                 ...   
18188     Das Boot - Die komplette TV-Serie     tv          Drama, War   
18194              The Amazon Review Killer     tv         Documentary   
18198                          Conspirators     tv             Mystery   
18200                   Extraordinary World     tv         Documentary   
18203  They Took Our Child: We Got Her Back     tv  Crime, Documentary   

       Release Year     IMDb ID  IMDb Average Rating       IMDb Num Votes  \
0              1995   tt0113101                  6.7               113403   
1              1994   tt0109830                  8.8              2348885   
2              1999   tt0169547                  8.3              1238903   
3              1941   tt0033467                  8.3               477481   
4              1927   tt0017136                  8.3               192285   
...             ...         ...                  ...                  ...   
18188          1985  tt30970892                  8.8                  198   
18194          2024  tt33342317                  6.4                   64   
18198          2025  tt35075008                  5.5                   44   
18200          2025  tt34888944                  NaN -9223372036854775808   
18203          2015   tt5056408                  8.6                   31   

      Service Name  
0          AppleTV  
1          AppleTV  
2          AppleTV  
3          AppleTV  
4          AppleTV  
...            ...  
18188      AppleTV  
18194      AppleTV  
18198      AppleTV  
18200      AppleTV  
18203      AppleTV  

[17641 rows x 8 columns]

In [15]:
#apple.genres.value_counts()
# apple['genres'].str.split(',', expand=True)
apple[["Genre 1","Genre 2","Genre 3","Genre 4","Genre 5","Genre 6"]] = apple["Genres"].str.split(',',expand=True)

display(apple)



Title   Type              Genres  \
0                                Four Rooms  movie              Comedy   
1                              Forrest Gump  movie      Drama, Romance   
2                           American Beauty  movie               Drama   
3                              Citizen Kane  movie      Drama, Mystery   
4                                Metropolis  movie       Drama, Sci-Fi   
...                                     ...    ...                 ...   
18188     Das Boot - Die komplette TV-Serie     tv          Drama, War   
18194              The Amazon Review Killer     tv         Documentary   
18198                          Conspirators     tv             Mystery   
18200                   Extraordinary World     tv         Documentary   
18203  They Took Our Child: We Got Her Back     tv  Crime, Documentary   

       Release Year     IMDb ID  IMDb Average Rating       IMDb Num Votes  \
0              1995   tt0113101                  6.7               113403   
1              1994   tt0109830                  8.8              2348885   
2              1999   tt0169547                  8.3              1238903   
3              1941   tt0033467                  8.3               477481   
4              1927   tt0017136                  8.3               192285   
...             ...         ...                  ...                  ...   
18188          1985  tt30970892                  8.8                  198   
18194          2024  tt33342317                  6.4                   64   
18198          2025  tt35075008                  5.5                   44   
18200          2025  tt34888944                  NaN -9223372036854775808   
18203          2015   tt5056408                  8.6                   31   

      Service Name      Genre 1       Genre 2 Genre 3 Genre 4 Genre 5 Genre 6  
0          AppleTV       Comedy          None    None    None    None    None  
1          AppleTV        Drama       Romance    None    None    None    None  
2          AppleTV        Drama          None    None    None    None    None  
3          AppleTV        Drama       Mystery    None    None    None    None  
4          AppleTV        Drama        Sci-Fi    None    None    None    None  
...            ...          ...           ...     ...     ...     ...     ...  
18188      AppleTV        Drama           War    None    None    None    None  
18194      AppleTV  Documentary          None    None    None    None    None  
18198      AppleTV      Mystery          None    None    None    None    None  
18200      AppleTV  Documentary          None    None    None    None    None  
18203      AppleTV        Crime   Documentary    None    None    None    None  

[17643 rows x 14 columns]

### **Hulu**

In [16]:
# View top 5 rows in the hulu df
hulu.head()

title   type                     genres  releaseYear  \
0                Ariel  movie     Comedy, Crime, Romance       1988.0   
1  Shadows in Paradise  movie       Comedy, Drama, Music       1986.0   
2         Forrest Gump  movie             Drama, Romance       1994.0   
3    The Fifth Element  movie  Action, Adventure, Sci-Fi       1997.0   
4   My Life Without Me  movie             Drama, Romance       2003.0   

      imdbId  imdbAverageRating  imdbNumVotes availableCountries  
0  tt0094675                7.4        8949.0                NaN  
1  tt0092149                7.5        7727.0                NaN  
2  tt0109830                8.8     2348885.0                NaN  
3  tt0119116                7.6      521815.0                NaN  
4  tt0314412                7.4       26140.0                NaN

In [17]:
hulu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10203 entries, 0 to 10202
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               9552 non-null   object 
 1   type                10203 non-null  object 
 2   genres              9853 non-null   object 
 3   releaseYear         10166 non-null  float64
 4   imdbId              9138 non-null   object 
 5   imdbAverageRating   8830 non-null   float64
 6   imdbNumVotes        8830 non-null   float64
 7   availableCountries  44 non-null     object 
dtypes: float64(3), object(5)
memory usage: 637.8+ KB


In [18]:
hulu.describe()

releaseYear  imdbAverageRating  imdbNumVotes
count  10166.000000        8830.000000  8.830000e+03
mean    2011.744344           6.569921  3.536248e+04
std       13.952201           1.055329  1.456587e+05
min     1929.000000           1.800000  5.000000e+00
25%     2009.000000           6.000000  1.390000e+02
50%     2016.000000           6.700000  9.210000e+02
75%     2021.000000           7.300000  7.385750e+03
max     2025.000000           9.500000  3.002274e+06

In [19]:
hulu.select_dtypes("object").describe()

title   type genres     imdbId availableCountries
count   9552  10203   9853       9138                 44
unique  9342      2    664       9134                  3
top     Prey  movie  Drama  tt1945851                 JP
freq       4   6019   1031          2                 28

In [20]:
# Drop availableCountries Column - no value with such a low non-null number
hulu.drop("availableCountries",axis=1,inplace=True)
hulu.columns

Index(['title', 'type', 'genres', 'releaseYear', 'imdbId', 'imdbAverageRating',
       'imdbNumVotes'],
      dtype='object')

In [21]:
# Change releaseYear,  imdbNumVotes to integers 
# Display only the data types

hulu[["releaseYear","imdbNumVotes"]] = hulu[["releaseYear","imdbNumVotes"]].apply(np.int64)
display(hulu.dtypes)

c:\Users\rkynhoff\OneDrive - University of Rochester\Documents\CodeLouisville\Streaming_Service_Comparisons\venv\Lib\site-packages\pandas\core\series.py:1031: RuntimeWarning: invalid value encountered in cast
  arr = np.asarray(values, dtype=dtype)


title                 object
type                  object
genres                object
releaseYear            int64
imdbId                object
imdbAverageRating    float64
imdbNumVotes           int64
dtype: object

In [22]:
# Rename Columns
hulu.rename(columns={"title":"Title","type":"Type","genres":"Genres","releaseYear":"Release Year","imdbId":"IMDb ID","imdbAverageRating":"IMDb Average Rating","imdbNumVotes":"IMDb Num Votes"},inplace = True)
# Add column that identifies the streaming service
hulu["Service Name"] = "Hulu"
hulu.columns

Index(['Title', 'Type', 'Genres', 'Release Year', 'IMDb ID',
       'IMDb Average Rating', 'IMDb Num Votes', 'Service Name'],
      dtype='object')

In [23]:
# Determine how many titles contain null values
hulu["Title"].isna().sum()

np.int64(651)

In [24]:
# Drop the rows contiaining null values in the Title column
hulu = hulu.dropna(subset=["Title"])
hulu["Title"].isna().sum()

np.int64(0)

In [25]:
hulu.duplicated().value_counts()

False    9550
True        2
Name: count, dtype: int64

In [26]:
hulu[hulu.duplicated(keep=False)]

Title Type              Genres  Release Year    IMDb ID  \
6811       The Gutsy Frog   tv              Comedy          2015  tt4658636   
7193  Scorned: Love Kills   tv  Crime, Documentary          2012  tt2287041   
7194  Scorned: Love Kills   tv  Crime, Documentary          2012  tt2287041   
8646       The Gutsy Frog   tv              Comedy          2015  tt4658636   

      IMDb Average Rating  IMDb Num Votes Service Name  
6811                  7.1              17         Hulu  
7193                  6.9             304         Hulu  
7194                  6.9             304         Hulu  
8646                  7.1              17         Hulu

In [27]:
hulu.drop_duplicates()


Title   Type  \
0                                     Ariel  movie   
1                       Shadows in Paradise  movie   
2                              Forrest Gump  movie   
3                         The Fifth Element  movie   
4                        My Life Without Me  movie   
...                                     ...    ...   
10194                          Scam Goddess     tv   
10196             AQUARION Myth of Emotions     tv   
10199  They Took Our Child: We Got Her Back     tv   
10201     Cardcaptor Sakura: Clear Card Arc     tv   
10202                        Faces of Music     tv   

                             Genres  Release Year     IMDb ID  \
0            Comedy, Crime, Romance          1988   tt0094675   
1              Comedy, Drama, Music          1986   tt0092149   
2                    Drama, Romance          1994   tt0109830   
3         Action, Adventure, Sci-Fi          1997   tt0119116   
4                    Drama, Romance          2003   tt0314412   
...                             ...           ...         ...   
10194              Crime, Talk-Show          2025  tt32821425   
10196    Action, Animation, Fantasy          2024  tt30644518   
10199            Crime, Documentary          2015   tt5056408   
10201  Adventure, Animation, Comedy          2018   tt6279576   
10202                   Documentary          2025  tt35460841   

       IMDb Average Rating       IMDb Num Votes Service Name  
0                      7.4                 8949         Hulu  
1                      7.5                 7727         Hulu  
2                      8.8              2348885         Hulu  
3                      7.6               521815         Hulu  
4                      7.4                26140         Hulu  
...                    ...                  ...          ...  
10194                  6.7                   58         Hulu  
10196                  3.5                   17         Hulu  
10199                  8.6                   31         Hulu  
10201                  7.7                  879         Hulu  
10202                  NaN -9223372036854775808         Hulu  

[9550 rows x 8 columns]

In [ ]:
#Obtain max number of elements in the Genres column
hulu["Genres"].str.split(", ",expand=True)


0            1        2     3     4     5     6
0           Comedy        Crime  Romance  None  None  None  None
1           Comedy        Drama    Music  None  None  None  None
2            Drama      Romance     None  None  None  None  None
3           Action    Adventure   Sci-Fi  None  None  None  None
4            Drama      Romance     None  None  None  None  None
...            ...          ...      ...   ...   ...   ...   ...
10194        Crime    Talk-Show     None  None  None  None  None
10196       Action    Animation  Fantasy  None  None  None  None
10199        Crime  Documentary     None  None  None  None  None
10201    Adventure    Animation   Comedy  None  None  None  None
10202  Documentary         None     None  None  None  None  None

[9552 rows x 7 columns]

In [35]:
# Split the elements in the Genres column while creating new columns for the max number 
hulu[["Genre 1","Genre 2","Genre 3","Genre 4","Genre 5","Genre 6","Genre 7"]] = hulu["Genres"].str.split(',',expand=True)

apple.head(2)

Title   Type          Genres  Release Year    IMDb ID  \
0    Four Rooms  movie          Comedy          1995  tt0113101   
1  Forrest Gump  movie  Drama, Romance          1994  tt0109830   

   IMDb Average Rating  IMDb Num Votes Service Name Genre 1   Genre 2 Genre 3  \
0                  6.7          113403      AppleTV  Comedy      None    None   
1                  8.8         2348885      AppleTV   Drama   Romance    None   

  Genre 4 Genre 5 Genre 6  
0    None    None    None  
1    None    None    None

In [ ]:
# View top 5 rows in the netflix df
netflix.head()

In [ ]:
# Add a column to netflix DF and entered a default value for each row & display bottom two rows
netflix["Service Name"] = "Netflix"
netflix.tail(2)

In [ ]:
# View top 5 rows of prime df
prime.head()

In [ ]:
# Add a column to the prime df and entered a default value for each row & display bottom two rows
prime["Service Name"] = "Prime"
prime.tail(2)

# map function for creating column and populating each cell within each df with the df name

# create a dictionary with the list of the names of df and have it look at the df name to populate the column

In [ ]:
# Merge the dataframes to append them to the end of each other since using the same column names
streaming = pd.concat([apple, hulu, netflix, prime], axis = 0)


In [ ]:
streaming.info()

In [ ]:
streaming.describe()

In [ ]:
streaming.duplicated().value_counts()

In [ ]:
streaming[streaming.duplicated(keep=False)]

In [ ]:
streaming["title"].isnull().value_counts()


In [ ]:
streaming["imdbId"].isnull().value_counts()

In [ ]:
# # Assign dataset names & combine to read in as separate data frames

# list_of_names = ['Netflix','Hulu','Prime','AppleTV']

# # Created an empty list into which I can place the datasets
# combined_list = []

# # Used a function to append the datasets into the empty list I created above
# for i in range(len(list_of_names)):
#     temp_df = pd.read_csv(list_of_names[i]+".csv")
#     combined_list.append(temp_df)

### **Initial Checks**

In [ ]:
# # Looking at file content of list in index 0 - Netflix - to determine column names, non-null counts, datatypes, and what the top five rows look like
# combined_list[0].info()
# combined_list[0].head(5)



In [ ]:
# # Looking at file content of list in index 1 - Hulu - to determine column names, non-null counts, datatypes, and what the top five rows look like
# combined_list[1].info()
# combined_list[1].head(5)

In [ ]:
# # Looking at file content of list in index 2 - Prime - to determine column names, non-null counts, datatypes, and what the top five rows look like
# combined_list[2].info()
# combined_list[2].head(5)

In [ ]:
# # Looking at file content of list in index 3 - AppleTV - to determine column names, non-null counts, datatypes, and what the top five rows look like
# combined_list[3].info()
# combined_list[3].head(5)

### Initiall Review (02/10/2025)
- The "availableCountries" column will not provide much data going forward
- No dataset has a column for the source of the data, so this will need to be added

### Secondary Review (02/12/2025)
- Each dataset contains an "imdbId" column with the listing's IMDb ID, which is a string; this may come in handy 
- There are 3,514 missing titles
    - Prime = 2234
    - AppleTV = 651
    - Hulu = 2885
    - Netflix = 629
- There are 2,590 duplicated rows, among which many are from missing titles, though several are duplicate title names 
    - Apple = 2
    - Hulu = 651


### Breakdown:
#### Apple
- 565 Missing Titles
- 2 Dupliate Rows
#### Hulu:
- 651 Missing Titles
- 2 Duplicated Rows